diff -ru SVGATextMode-1.10.orig/XFREE/Makefile SVGATextMode-1.10/XFREE/Makefile
--- SVGATextMode-1.10.orig/XFREE/Makefile	2004-05-26 19:00:45.081665440 +0200
+++ SVGATextMode-1.10/XFREE/Makefile	2004-05-26 19:01:50.284753064 +0200
@@ -22,11 +22,11 @@
           mach64/mach64_mem_access.o mach64/mach64.o \
           riva128_clock.o \
           voodoo_clock.o \
-          xfree_compat.o
+          xfree_compat.o \
+	  common_hw/IODelay.o
 
 #          mach64/mach64clockchips.o \
 
-ASMOBJS = common_hw/IODelay.o
 
 
 all: $(XFREELIBS) $(OBJECTS) $(ASMOBJS)
diff -ru SVGATextMode-1.10.orig/vga_prg.h SVGATextMode-1.10/vga_prg.h
--- SVGATextMode-1.10.orig/vga_prg.h	2004-05-26 18:55:31.629317440 +0200
+++ SVGATextMode-1.10/vga_prg.h	2004-05-26 19:01:50.284753064 +0200
@@ -41,6 +41,26 @@
 #  define iopl(x) (0)
 #else
 #  include <sys/io.h>
+#  ifdef __alpha__
+#    ifndef inb_p
+#      define  inb_p   inb
+#    endif
+#    ifndef inw_p
+#      define  inw_p   inw
+#    endif
+#    ifndef inl_p
+#      define  inl_p   inl
+#    endif
+#    ifndef outb_p
+#      define  outb_p  outb
+#    endif
+#    ifndef outw_p
+#      define  outw_p  outw
+#    endif
+#    ifndef outl_p
+#      define  outl_p  outl
+#    endif
+#  endif /* __alpha__ */
 #endif
 
 #ifndef _SYS_IO_H
